File: Day10/main.ts
function compile (instructions: string[]): number {
  const registers = {};
    let pointer = 0;

    while (pointer < instructions.length) {
        const [command, arg1, arg2] = instructions[pointer].split(' ');

        if (command === 'MOV') registers[arg2] = isNaN(arg1) ? (registers[arg1] || 0) : +arg1;
        else if (command === 'INC') registers[arg1] = (registers[arg1] || 0) + 1;
        else if (command === 'DEC') registers[arg1] = (registers[arg1] || 0) - 1;
        else if (command === 'JMP' && !(registers[arg1] || 0)) {
            pointer = +arg2;
            continue;
        }

        pointer++;
    }

    return registers.A;
}

File: Day10/main.py
def compile(instructions):
    registers = {}
    pointer = 0

    while pointer < len(instructions):
        parts = instructions[pointer].split()
        command, *args = parts

        if command == 'MOV':
            registers[args[1]] = int(args[0]) if args[0].isdigit() else registers.get(args[0], 0)
        elif command == 'INC':
            registers[args[0]] = registers.get(args[0], 0) + 1
        elif command == 'DEC':
            registers[args[0]] = registers.get(args[0], 0) - 1
        elif command == 'JMP':
            if registers.get(args[0], 0) == 0:
                pointer = int(args[1])
                continue

        pointer += 1

    return registers.get('A')
